sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala

```diff
@@ -843,133 +987,167 @@ class GpuMergeAggregateIterator(
-            }
-          }
-          return
+  private lazy val concatAndMergeHelper =
+    new AggHelper(inputAttributes, groupingExpressions, aggregateExpressions,
+      forceMerge = true, useTieredProject = useTieredProject)
+
+  private def cbIteratorStealingFromBuffer(input: ListBuffer[SpillableColumnarBatch]) = {
+    val aggregatedBatchIter = new Iterator[ColumnarBatch] {
+      override def hasNext: Boolean = input.nonEmpty
+
+      override def next(): ColumnarBatch = {
+        withResource(input.remove(0)) { spillable =>
+          spillable.getColumnarBatch()
         }
       }
     }
+    aggregatedBatchIter
   }
 
-  /**
-   * Perform a single pass over the aggregated batches attempting to merge adjacent batches.
-   * @return true if at least one merge operation occurred
-   */
-  private def mergePass(): Boolean = {
-    val batchesToConcat: mutable.ArrayBuffer[SpillableColumnarBatch] = mutable.ArrayBuffer.empty
-    var wasBatchMerged = false
-    // Current size in bytes of the batches targeted for the next concatenation
-    var concatSize: Long = 0L
-    var batchesLeftInPass = aggregatedBatches.size()
+  private case class RepartitionAggregateIterator(
+      inputBatches: ListBuffer[SpillableColumnarBatch],
+      hashKeys: Seq[GpuExpression],
+      targetSize: Long,
+      opTime: GpuMetric,
+      repartitionTime: GpuMetric) extends Iterator[ColumnarBatch]
+    with AutoCloseable {
 
-    while (batchesLeftInPass > 0) {
-      closeOnExcept(batchesToConcat) { _ =>
-        var isConcatSearchFinished = false
-        // Old batches are picked up at the front of the queue and freshly merged batches are
-        // appended to the back of the queue. Although tempting to allow the pass to "wrap around"
-        // and pick up batches freshly merged in this pass, it's avoided to prevent changing the
-        // order of aggregated batches.
-        while (batchesLeftInPass > 0 && !isConcatSearchFinished) {
-          val candidate = aggregatedBatches.getFirst
-          val potentialSize = concatSize + candidate.sizeInBytes
-          isConcatSearchFinished = concatSize > 0 && potentialSize > targetMergeBatchSize
-          if (!isConcatSearchFinished) {
-            batchesLeftInPass -= 1
-            batchesToConcat += aggregatedBatches.removeFirst()
-            concatSize = potentialSize
+    case class AggregatePartition(batches: ListBuffer[SpillableColumnarBatch], seed: Int)
+      extends AutoCloseable {
+      override def close(): Unit = {
+        batches.safeClose()
+      }
+
+      def totalRows(): Long = batches.map(_.numRows()).sum
+
+      def totalSize(): Long = batches.map(_.sizeInBytes).sum
+
+      def isAllBatchesSingleRow: Boolean = {
+        batches.forall(_.numRows() == 1)
+      }
+
+      def split(): ListBuffer[AggregatePartition] = {
+        withResource(new NvtxWithMetrics("agg repartition", NvtxColor.CYAN, repartitionTime)) { _ =>
+          if (seed >= hashSeed + 100) {
```

Comment:
Why would we ever need to repartition the data more than once?

The current code does a single aggregation pass through the data. Once you have done that pass you know the statistics about the data and should be able to make a very good guess about how to combine the data based on the number of shuffle partitions. Is this because there might be a large number of hash collisions?  I think in practice that would never happen, but I would like to understand the reasoning here.